shap/explainers/linear.py

```diff
@@ -90,11 +90,18 @@ def __init__(self, model, data, nsamples=1000, feature_dependence=None):
         # Note: mean can be numpy.matrixlib.defmatrix.matrix or numpy.matrix type depending on numpy version
         if sp.sparse.issparse(self.mean) or str(type(self.mean)).endswith("matrix'>"):
             # accept both sparse and dense coef
-            if not sp.sparse.issparse(self.coef):
-                self.coef = np.asmatrix(self.coef)
-            self.expected_value = self.coef.dot(self.mean.T) + self.intercept
+            # if not sp.sparse.issparse(self.coef):
+            #     self.coef = np.asmatrix(self.coef)
+            self.expected_value = np.dot(self.coef, self.mean) + self.intercept
```

Comment:
I don't think np.dot works for sparse mean, but I'm not sure if the mean will ever be sparse (I think it can be if most of the columns are all-zero), I will need to double check that

Reply:
according to this documentation it is always matrix:
https://docs.scipy.org/doc/scipy-0.19.0/reference/generated/scipy.sparse.csr_matrix.mean.html

Reply:
I ran some tests against this branch from my test suite and I am getting a shapes not aligned error when doing np.dot, "shapes (58081,) and (1,58081) not aligned", I think self.mean is 2-D while self.coef is 1-D in this case which causes the error:

  File "C:\azuremlcli\tests\scenarios\explain_model\test_mimic_explainer.py", line 99, in test_explain_model_npz_linear
    verifier.verify_explain_model_npz_linear()
  File "C:\azuremlcli\tests\scenarios\explain_model\common_tabular_tests.py", line 440, in verify_explain_model_npz_linear
    explainer.explain_global(x_test, **policy)
  File "c:\azuremlcli\src\azureml-explain-model\azureml\explain\model\mimic\mimic_explainer.py", line 270, in explain_global
    batch_size=batch_size)
  File "c:\azuremlcli\src\azureml-explain-model\azureml\explain\model\mimic\mimic_explainer.py", line 214, in _get_explain_global_kwargs
    local_explanation = self.explain_local(evaluation_examples)
  File "c:\azuremlcli\src\azureml-explain-model\azureml\explain\model\dataset\decorator.py", line 22, in explain_func_wrapper
    return explain_func(self, evaluation_examples, *args, **kwargs)
  File "c:\azuremlcli\src\azureml-explain-model\azureml\explain\model\mimic\mimic_explainer.py", line 350, in explain_local
    kwargs = self._get_explain_local_kwargs(evaluation_examples)
  File "c:\azuremlcli\src\azureml-explain-model\azureml\explain\model\mimic\mimic_explainer.py", line 300, in _get_explain_local_kwargs
    local_importance_values = self.surrogate_model.explain_local(dataset, **kwargs)
  File "c:\azuremlcli\src\azureml-explain-model\azureml\explain\model\mimic\models\linear_model.py", line 192, in explain_local
    self.covariance, self.random_state)
  File "c:\azuremlcli\src\azureml-explain-model\azureml\explain\model\mimic\models\linear_model.py", line 55, in _create_linear_explainer
    return shap.LinearExplainer((model_coef, model_intercept), (mean, covariance))
  File "c:\shap\shap\explainers\linear.py", line 95, in __init__
    self.expected_value = np.dot(self.coef, self.mean) + self.intercept
ValueError: shapes (58081,) and (1,58081) not aligned: 58081 (dim 0) != 1 (dim 0)

Reply:
Thanks! I'll update that then

Reply:
nevermind, it looks like I am passing the mean as a matrix to LinearExplainer.  I should change it to numpy array.

Reply:
I've confirmed that once I changed the mean to numpy array my tests passed, so the changes in this PR look good

Reply:
Great, thanks!